delete-split-update-make_chromarms

gfudenberg opened the pull request:
- deleted split, and ``make_chromarms`` now depends on ``subtract``
- could use a few more tests for ``make_chromarms``
- streamlined subtract such that it doesn't loop over groups, since ``complement`` already iterates over groups. *however*, ``np.iinfo(np.int64).max`` needs to be cast to ``pd.Int64Dtype()`` otherwise it will get changed to a float with ``overlap(how='left')``... 
- added ``return_index`` to ``subtract()`` that returns: the original index, and the sub-index. The sub-index counts how many intervals an original interval is divided into. This is conveniently passed to a dictionary in ``make_chromarms`` to add "_p" or "_q" suffixes.
- we may want to change the input checking / flexility of input types for ``make_chromarms``.

gfudenberg commented on bioframe/extras.py:
```diff
@@ -20,7 +20,7 @@ def make_chromarms(
     cols_chroms=("chrom", "length"),
     cols_mids=("chrom", "mid"),
-    suffixes=("_p", "_q"),
+    sub_index_to_suffix={0: "_p", 1: "_q"},
```
do you mean dict?

golobor replied:
A tuple, or, a list, yeah. It's just replacing a tuple with a dict that is indexed by sequential integers is weird ...

gfudenberg replied:
okok, can roll that back

gfudenberg commented on bioframe/ops.py:
```diff
@@ -1114,32 +1131,36 @@ def subtract(df1, df2, suffixes=("", "_"), cols1=None, cols2=None):
         name_updates[i + suffixes[0]] = i
+    if return_index:
+        name_updates["index" + suffixes[0]] = "index" + suffixes[0]
+        name_updates["index" + suffixes[1]] = "complement_index" + suffixes[1]
```
it gets dropped later-- its just converted to sub_index

gfudenberg commented:
quick update:  I updated `is_bedframe` & its docstring as per our discussion on 7/1. This adds a NA checking step that requires for each each interval, if any of chrom, start, end are NA, then all are NA. I updated select, trim, and expand to be NA friendly & added tests with NAs for those functions. 

I was thinking about how to handle NA in `overlap()` and came across this interesting quirk of groupby:
![image](https://user-images.githubusercontent.com/19657573/124425423-5d0f3580-dd1d-11eb-9ee9-492a9ec9554c.png)

also, I realized that the keep_order=True only really makes sense with how="left". For example, it re-orders to put right intervals without overlaps first with how="outer". 

One (not particularly elegant) possibility is to take out all intervals that have a NA in anything that is getting a groupby([chrom]+[on]) before passing to `_overlap_intidxs` and then add them back in afterwards... 

other suggestions on how to handle all this?

golobor commented:
yeah, i guess, indeed, keep_order indeed only makes sense when how='inner'/'left'?.. :( I never really thought of this. I guess, that's ok, the users can swap df1 and df2, after all. I'd say, at this point, we just need to mention this in the docstring?

golobor commented:
> One (not particularly elegant) possibility is to take out all intervals that have a NA in anything that is getting a groupby([chrom]+[on]) before passing to _overlap_intidxs and then add them back in afterwards...

Yeah, actually, this may be the most robust solution, no?..

gfudenberg commented:
> yeah, i guess, indeed, keep_order indeed only makes sense when how='inner'/'left'?.. :( I never really thought of this. I guess, that's ok, the users can swap df1 and df2, after all. I'd say, at this point, we just need to mention this in the docstring?

Or make `keep_order=True` incompatible with how='right'/'outer'  (i.e. raise an error)?

golobor commented:
Oh , right, this also sounds right - an error and a docstring.

On Fri, Jul 9, 2021, 23:48 gfudenberg ***@***.***> wrote:

> yeah, i guess, indeed, keep_order indeed only makes sense when
> how='inner'/'left'?.. :( I never really thought of this. I guess, that's
> ok, the users can swap df1 and df2, after all. I'd say, at this point, we
> just need to mention this in the docstring?
>
> Or make keep_order=True incompatible with how='right'/'outer' (i.e. raise
> an error)?
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/open2c/bioframe/pull/83#issuecomment-877473545>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAG64CVHJEYZMAXZI5I6GLTTW5VBNANCNFSM466ODAIQ>
> .
>

gfudenberg commented:
working on making `merge` NA friendly, ran into this 'fun' behavior of pd.concat:

![image](https://user-images.githubusercontent.com/19657573/125176857-e6928d80-e18b-11eb-9210-98fa444b9d54.png)

gfudenberg commented:
a couple more questions/thoughts:
- trim & sort_bedframe have different syntax for inferring `df_view_col` & should probably be synchronized. Any preference on the syntax in sort_bedframe(passing: `infer_assignment=True, df_view_col='view_region'`)  or trim (`df_view_col=None`)?
- `pair_by_distance` should probably: 
  - make use of `bioframe.sort_bedframe`
  - not allow negative `min_sep`
  - rename `from_ends` to 'measurement_position` and make it a string ('ends', 'mids') rather than a bool
  - move out of `bioframe.ops` into `bioframe.extras`

golobor commented:
> a couple more questions/thoughts:
> 
> * trim & sort_bedframe have different syntax for inferring `df_view_col` & should probably be synchronized. Any preference on the syntax in sort_bedframe(passing: `infer_assignment=True, df_view_col='view_region'`)  or trim (`df_view_col=None`)?

I like the 2nd option (a-la trim) much more!

> * `pair_by_distance` should probably:
>   
>   * make use of `bioframe.sort_bedframe`
>   * not allow negative `min_sep`

yes!

>   * rename `from_ends` to 'measurement_position` and make it a string ('ends', 'mids') rather than a bool

agreed, too. `measurement_position` is a bit bulky of a name? "relative_to"? "distance_between"? 

>   * move out of `bioframe.ops` into `bioframe.extras`

Yeah, I agree, it's not that "kernel".

> also note that `coverage` and `count_overlaps` do not have the `keep_order` option (see commit above), as the groupby indicies sorted the dataFrame anyways.

gfudenberg commented:
thanks for the suggestions! updated.

gfudenberg commented:
update to make_viewframe & trim should close https://github.com/open2c/bioframe/issues/80 & https://github.com/open2c/bioframe/issues/87

Phlya commented:
Is anything stopping us from merging this and releasing?

gfudenberg commented:
nope! bumped the version & was planning to go through issues and close some out next
